File: microservices/auth-service/src/main/java/com/karate/authservice/infrastructure/messaging/UserDeletedListener.java
package com.karate.authservice.infrastructure.messaging;

import com.karate.authservice.domain.repository.AuthUserRepository;
import com.karate.authservice.infrastructure.messaging.dto.UserDeletedEvent;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.stereotype.Component;

@Slf4j
@Component
@RequiredArgsConstructor
public class UserDeletedListener {

    private final AuthUserRepository authUserRepository;


    @KafkaListener(topics = "${topics.user-deleted}", groupId = "auth-service-user-deleted")
    @Transactional
    public void onUserDeleted(UserDeletedEvent evt) {
        Long userId = evt.userId();
        log.info("auth-service: UserDeletedEvent userId={}", userId);
        authUserRepository.deleteByUserId(userId);
    }
}

File: microservices/feedback-service/src/main/kotlin/com/karate/feedback_service/infrastructure/messaging/dto/UserDeletedListener.kt
package com.karate.feedback_service.infrastructure.messaging.dto

import com.karate.feedback_service.domain.repository.FeedbackRepository
import org.slf4j.LoggerFactory
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class UserDeletedListener(
    private val feedbackRepository: FeedbackRepository
) {

    private val log = LoggerFactory.getLogger(UserDeletedListener::class.java)

    @Transactional
    @KafkaListener(topics = ["user.deleted"], groupId = "feedback-service-user-deleted")
    fun onUserDeleted(evt: UserDeletedEvent) {
        val count = feedbackRepository.deleteAllByUserId(evt.userId)
        log.info("feedback-service: handled USER_DELETED userId={} deleted={} feedback(s)", evt.userId, count)
    }
}